### ISSUE TITLE
Tracking Issue for Flaky File-like obj CI failure

### ISSUE THREAD

**Issue by mthrok** [pytorch/audio#1229]
Some tests for file-like object support for `load` function are failing sporadicly. Retrying the job makes the failure go away, but we need to understand what is causing the flaky failure.

**Comment by mthrok** [comment]
[Example job failing](https://app.circleci.com/pipelines/github/pytorch/audio/4801/workflows/1d1894d9-fcf8-4ce3-8530-cba2d519a053/jobs/154578) for PR #1281

Observations:
 - All the tests are `flac`.
   - `test_bytesio_3_flac`
   - `test_bytesio_4_flac`
   - `test_bytesio_5_flac`
   - `test_fileobj_3_flac`
   - `test_fileobj_4_flac`
   - `test_fileobj_5_flac`

CPU Info:
```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 85
model name	: Intel(R) Xeon(R) Platinum 8124M CPU @ 3.00GHz
stepping	: 4
microcode	: 0x2006906
cpu MHz		: 3402.471
cache size	: 25344 KB
physical id	: 0
siblings	: 36
core id		: 0
cpu cores	: 18
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa itlb_multihit
bogomips	: 5999.99
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:
```

**Comment by mthrok** [comment]
Job failing on [nightly unittest](https://app.circleci.com/pipelines/github/pytorch/audio/4806/workflows/a673db64-0412-48ff-a38e-44120c9b5fe9/jobs/154790)

Failed tests:
- `test_bytesio_3_flac`
- `test_bytesio_4_flac`
- `test_bytesio_5_flac`
- `test_fileobj_3_flac`
- `test_fileobj_4_flac`
- `test_fileobj_5_flac`

Configuration
- `gpu_py3.6`
- CPU Info    
```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
stepping	: 1
microcode	: 0xb000038
cpu MHz		: 2699.984
cache size	: 46080 KB
physical id	: 0
siblings	: 16
core id		: 0
cpu cores	: 8
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx xsaveopt
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs
bogomips	: 4600.14
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:
```

CPU Info for [successful job: `gpu_py3.8`](https://app.circleci.com/pipelines/github/pytorch/audio/4806/workflows/a673db64-0412-48ff-a38e-44120c9b5fe9/jobs/154799)

```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
stepping	: 1
microcode	: 0xb000038
cpu MHz		: 2699.894
cache size	: 46080 KB
physical id	: 0
siblings	: 16
core id		: 0
cpu cores	: 8
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx xsaveopt
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs
bogomips	: 4600.07
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:
```

**Comment by mthrok** [comment]
Maybe this might be the cause;


```
[5/7] c++ -MMD -MF /scratch/moto/torchaudio/build/temp.linux-x86_64-3.8/scratch/moto/torchaudio/torchaudio/csrc/sox/effects_chain.o.d -pthread -B /home/moto/conda/envs/PY3.8-cuda101/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/scratch/moto/torchaudio -I/scratch/moto/torchaudio/third_party/install/include -I/home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/include -I/home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -I/home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/include/TH -I/home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/include/THC -I/home/moto/conda/envs/PY3.8-cuda101/include/python3.8 -c -c /scratch/moto/torchaudio/torchaudio/csrc/sox/effects_chain.cpp -o /scratch/moto/torchaudio/build/temp.linux-x86_64-3.8/scratch/moto/torchaudio/torchaudio/csrc/sox/effects_chain.o -O3 -DBUILD_TRANSDUCER -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=_torchaudio -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
In file included from /home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/include/ATen/Parallel.h:140:0,
                 from /home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/include/torch/csrc/api/include/torch/utils.h:3,
                 from /home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/include/torch/csrc/api/include/torch/nn/cloneable.h:5,
                 from /home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/include/torch/csrc/api/include/torch/nn.h:3,
                 from /home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/include/torch/csrc/api/include/torch/all.h:13,
                 from /home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/include/torch/extension.h:4,
                 from /scratch/moto/torchaudio/torchaudio/csrc/sox/utils.h:8,
                 from /scratch/moto/torchaudio/torchaudio/csrc/sox/effects_chain.h:5,
                 from /scratch/moto/torchaudio/torchaudio/csrc/sox/effects_chain.cpp:1:
/home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/include/ATen/ParallelOpenMP.h:83:0: warning: ignoring #pragma omp parallel [-Wunknown-pragmas]
 #pragma omp parallel for if ((end - begin) >= grain_size)

/scratch/moto/torchaudio/torchaudio/csrc/sox/effects_chain.cpp: In function ‘int torchaudio::sox_effects_chain::{anonymous}::fileobj_input_drain(sox_effect_t*, sox_sample_t*, size_t*)’:
/scratch/moto/torchaudio/torchaudio/csrc/sox/effects_chain.cpp:357:20: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   if (num_consumed > priv->buffer_size) {
       ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
/scratch/moto/torchaudio/torchaudio/csrc/sox/effects_chain.cpp:369:20: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     if (num_refill < num_consumed) {
         ~~~~~~~~~~~^~~~~~~~~~~~~~
```

**Comment by mthrok** [comment]
The issue also happened on my dev env with the following command

```
$ BUILD_SOX=1 python setup.py develop && (cd test && pytest torchaudio_unittest/sox_io_backend/save_test.py )
```

**Comment by mthrok** [comment]
[Job failing](https://app.circleci.com/pipelines/github/pytorch/audio/4844/workflows/ce32c74e-50e4-4573-b79a-9ddf42009598/jobs/156197) on #1187

Failed tests;
 - test_bytesio_4_mp3
 - test_bytesio_tiny_4_mp3
 - test_bytesio_tiny_6_vorbis
 - test_fileobj_4_mp3
 - test_tarfile_4_mp3
 - test_requests_4_mp3
 - test_bytesio_1_mp3
 - test_bytesio_2_mp3
 - test_bytesio_clogged_1_mp3
 - test_bytesio_clogged_2_mp3
 - test_fileobj_1_mp3
 - test_fileobj_2_mp3
 - test_tarfile_1_mp3
 - test_tarfile_2_mp3
 - test_requests_1_mp3
 
CPU Info:
```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
stepping	: 1
microcode	: 0xb000038
cpu MHz		: 2693.246
cache size	: 46080 KB
physical id	: 0
siblings	: 16
core id		: 0
cpu cores	: 8
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx xsaveopt
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs
bogomips	: 4600.01
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:
```

**Comment by mthrok** [comment]
So far, the failing tests are container formats and the reason is shape mismatch.
One hypothesis is that the when refiling the internal byte buffer, the end of the buffer align with the end of packet so that the decoding ends early on.

**Comment by mthrok** [comment]
On e5c4de8, reproduced the error on my dev env.

<details><summary>script</summary>

```bash
#/usr/bin/env bash

set -e

cd test

while true; do
    TORCHAUDIO_TEST_TEMP_DIR=../tmp/ pytest torchaudio_unittest/backend/sox_io > debug_fileobj.log 2>&1
done
```
</details> It takes a while to get the failing result.

Observed two failure modes.
1. `load_test` and `info_test` failing. [[log](https://gist.github.com/mthrok/6b997dd6681a8e525834c7d955e33adc)]
2. `save_test` failing. [[log](https://gist.github.com/mthrok/28ef1b0c022d752fb0b7273189bb97cf)]

For the failure mode 1, all the `info` tests are failing as **`info` can correctly inspect the meta data of mp3**. The test expect that `info` function to return `num_frames == 0`, but occasionally, it returns some non-zero values. 

```
        assert sinfo.num_channels == num_channels
>       assert sinfo.num_frames == num_frames
E       AssertionError: assert 1152 == 0
E        +  where 1152 = <torchaudio.backend.common.AudioMetaData object at 0x7fd974117fa0>.num_frames
```
TODO:
* Check if `1152` is a correct value.
* Check if the audio data generated with `sox` has some randomness.

For the failure mode 2, all the failing tests are FLAC format and all of them are due to the un-matching Tensor shape.

**Comment by mthrok** [comment]
Checking at the data from the failure mode 2; 

[torchaudio_unittest.backend.sox_io.save_test.SaveTest.test_save_flac_bytesio_16_0.tar.gz](https://github.com/pytorch/audio/files/6007083/torchaudio_unittest.backend.sox_io.save_test.SaveTest.test_save_flac_bytesio_16_0.tar.gz)

looks like ~the conversion from flac to wav is altering the data.~ flac data has something wrong.

```
$ soxi test/debug_fileobj/save/debug_fileobj.6/torchaudio_unittest.backend.sox_io.save_test.SaveTest.test_save_flac_bytesio_16_0/*

Input File     : 'test/debug_fileobj/save/debug_fileobj.6/torchaudio_unittest.backend.sox_io.save_test.SaveTest.test_save_flac_bytesio_16_0/1.source.wav'
Channels       : 2
Sample Rate    : 8000
Precision      : 32-bit
Duration       : 00:00:03.00 = 24000 samples ~ 225 CDDA sectors
File Size      : 192k
Bit Rate       : 512k
Sample Encoding: 32-bit Signed Integer PCM


Input File     : 'test/debug_fileobj/save/debug_fileobj.6/torchaudio_unittest.backend.sox_io.save_test.SaveTest.test_save_flac_bytesio_16_0/2.1.torchaudio.flac'
Channels       : 2
Sample Rate    : 8000
Precision      : 16-bit
Duration       : 00:00:03.00 = 24000 samples ~ 225 CDDA sectors
File Size      : 11.0k
Bit Rate       : 29.5k
Sample Encoding: 16-bit FLAC


Input File     : 'test/debug_fileobj/save/debug_fileobj.6/torchaudio_unittest.backend.sox_io.save_test.SaveTest.test_save_flac_bytesio_16_0/2.2.result.wav'
Channels       : 2
Sample Rate    : 8000
Precision      : 25-bit
Duration       : 00:00:02.88 = 23040 samples ~ 216 CDDA sectors
File Size      : 184k
Bit Rate       : 512k
Sample Encoding: 32-bit Floating Point PCM


Input File     : 'test/debug_fileobj/save/debug_fileobj.6/torchaudio_unittest.backend.sox_io.save_test.SaveTest.test_save_flac_bytesio_16_0/3.1.sox.flac'
Channels       : 2
Sample Rate    : 8000
Precision      : 16-bit
Duration       : 00:00:03.00 = 24000 samples ~ 225 CDDA sectors
File Size      : 11.5k
Bit Rate       : 30.8k
Sample Encoding: 16-bit FLAC
Comment        : 'Comment=Processed by SoX'


Input File     : 'test/debug_fileobj/save/debug_fileobj.6/torchaudio_unittest.backend.sox_io.save_test.SaveTest.test_save_flac_bytesio_16_0/3.2.ref.wav'
Channels       : 2
Sample Rate    : 8000
Precision      : 25-bit
Duration       : 00:00:03.00 = 24000 samples ~ 225 CDDA sectors
File Size      : 192k
Bit Rate       : 512k
Sample Encoding: 32-bit Floating Point PCM

Total Duration of 5 files: 00:00:14.88
```

Audacity also reports that the final part of data is missing.

<img width="1664" alt="Screen Shot 2021-02-18 at 20 49 25" src="https://user-images.githubusercontent.com/855818/108445685-cee15980-722a-11eb-9da2-fa152d418342.png">

**Comment by mthrok** [comment]
Checking the data from failure mode 2; the source audio file and audio files processed with `sox` have the same checksum. This indicates that there is some un-deterministic behavior in torchaudio's implementation.

## successful case
```
$ sha256sum tmp/torchaudio_unittest.backend.sox_io.save_test.SaveTest.test_save_flac_bytesio_16_0/*
1a43dd5202839020aa2bc8dbdf10bd5d4038359221f0bfd5a310233fa18f74a9  1.source.wav
94d012cd58ab008f214af795b95f36688a1fa6263df2bbfc8176f04f36280bf9  2.1.torchaudio.flac
7dee548df31d8e8f53d208568e5f559f6641ea8598bf464b5c1e0c23dd714462  2.2.result.wav
7471e954e0fd44a710b8ed75a42370ba2b709d63b47eb30789db13d02fc85033  3.1.sox.flac
7dee548df31d8e8f53d208568e5f559f6641ea8598bf464b5c1e0c23dd714462  3.2.ref.wav
```

## Failure case
```
$ sha256sum test/debug_fileobj/save/debug_fileobj.6/torchaudio_unittest.backend.sox_io.save_test.SaveTest.test_save_flac_bytesio_16_0/*
1a43dd5202839020aa2bc8dbdf10bd5d4038359221f0bfd5a310233fa18f74a9  1.source.wav
efa33771a51c6ccbb41615c2729335aa8f5a0782dc898e6599305493e2d5dd27  2.1.torchaudio.flac
d63b694fea46b766676589b1fb8ce720b495c22e3282dae0e830e5d9e6fe86ae  2.2.result.wav
7471e954e0fd44a710b8ed75a42370ba2b709d63b47eb30789db13d02fc85033  3.1.sox.flac
7dee548df31d8e8f53d208568e5f559f6641ea8598bf464b5c1e0c23dd714462  3.2.ref.wav
```

**Comment by mthrok** [comment]
I updated the debug wrapper to check the content of data flowing in/out of Python/C++. [code](https://github.com/mthrok/audio/blob/0e4f38af3ebdaa21e49d559ab39da5d1f24042ab/torchaudio/backend/sox_io_backend.py#L19-L36)

Using this, I checked what is happening to `info` function test. Currently the following tests can fail. ([full log](https://gist.github.com/mthrok/9946c413b5106cedc63d6e59b15dfdb2))

```
FAILED torchaudio_unittest/backend/sox_io/info_test.py::TestFileObject::test_bytesio_4_mp3
FAILED torchaudio_unittest/backend/sox_io/info_test.py::TestFileObject::test_bytesio_tiny_4_mp3
FAILED torchaudio_unittest/backend/sox_io/info_test.py::TestFileObject::test_bytesio_tiny_6_vorbis
FAILED torchaudio_unittest/backend/sox_io/info_test.py::TestFileObject::test_fileobj_4_mp3
FAILED torchaudio_unittest/backend/sox_io/info_test.py::TestFileObject::test_tarfile_4_mp3
```

This test reports that the following data with the MD5 check sum were passed to C++ code.

```
9bd4a08354309652df6fffab34e13f88 torchaudio_unittest.backend.sox_io.info_test.TestFileObject.test_bytesio_4_mp3/test.mp3 (4096 bytes)
9b5c38ffb7ac5ccb996d77430038b31f torchaudio_unittest.backend.sox_io.info_test.TestFileObject.test_bytesio_tiny_4_mp3/test.mp3 (649 bytes)
00acddd8a7ceaae3fa0175f0a44b9880 torchaudio_unittest.backend.sox_io.info_test.TestFileObject.test_bytesio_tiny_6_vorbis/test.vorbis (3106 bytes)
9bd4a08354309652df6fffab34e13f88 torchaudio_unittest.backend.sox_io.info_test.TestFileObject.test_fileobj_4_mp3/test.mp3 (4096 bytes)
9bd4a08354309652df6fffab34e13f88 torchaudio_unittest.backend.sox_io.info_test.TestFileObject.test_tarfile_4_mp3/test.mp3 (4096 bytes)
```

When I ran the same tests successfully the same set of data report that they have the same check sum.

```
$ head -c 4096 ./tmp/torchaudio_unittest.backend.sox_io.info_test.TestFileObject.test_bytesio_4_mp3/test.mp3 | md5sum
9bd4a08354309652df6fffab34e13f88  -
$ head -c 649 ./tmp/torchaudio_unittest.backend.sox_io.info_test.TestFileObject.test_bytesio_tiny_4_mp3/test.mp3 | md5sum
9b5c38ffb7ac5ccb996d77430038b31f
$ head -c 3106 ./tmp/torchaudio_unittest.backend.sox_io.info_test.TestFileObject.test_bytesio_tiny_6_vorbis/test.vorbis | md5sum
00acddd8a7ceaae3fa0175f0a44b9880
$ head -c 4096 ./tmp/torchaudio_unittest.backend.sox_io.info_test.TestFileObject.test_fileobj_4_mp3/test.mp3 | md5sum
9bd4a08354309652df6fffab34e13f88
$ head -c 4096 ./tmp/torchaudio_unittest.backend.sox_io.info_test.TestFileObject.test_tarfile_4_mp3/test.mp3 | md5sum
9bd4a08354309652df6fffab34e13f88
```

This indicates that even though the data we pass to C++ are same, the code is somehow generating the different result.

**Comment by mthrok** [comment]
Looking further into the behavior of sox in info test case, it seems that when `mp3` file is opened, the length information is already un-deterministic [here](https://github.com/dmkrepo/libsox/blob/b9dd1a86e71bbd62221904e3e59dfaa9e5e72046/src/formats.c#L578). 

log from failure case (`length` is filled properly)
```
Read: 9bd4a08354309652df6fffab34e13f88 4096 / 4096 (4096)
STARTREAD:
  length: 20736
```

log from success case (`length` is not filled as expected)
```
Read: 9bd4a08354309652df6fffab34e13f88 4096 / 4096 (4096)
STARTREAD:
  length: 0
```

**Comment by mthrok** [comment]
For mp3, looks like it is happening somewhere in

https://github.com/dmkrepo/libsox/blob/b9dd1a86e71bbd62221904e3e59dfaa9e5e72046/src/mp3.c#L381-L394

**Comment by mthrok** [comment]
Looks like when info test fails `ft->seekable=1` while, most of the time, `ft->seekable=0`

**Comment by mthrok** [comment]
`is_seekable` function sometimes returns 1.

https://github.com/dmkrepo/libsox/blob/b9dd1a86e71bbd62221904e3e59dfaa9e5e72046/src/formats.c#L329-L338

**Comment by mthrok** [comment]
Okay, I think I figured. The `is_seekable` function is checking the `stat.st_mode` of an opened file, which is invalid in case of memory object.

https://github.com/dmkrepo/libsox/blob/b9dd1a86e71bbd62221904e3e59dfaa9e5e72046/src/formats.c#L336

**Comment by mthrok** [comment]
Resolved by #1297
